Event loop shutdown failures on 0.17

We use the following to define a session scoped event loop (per https://github.com/pytest-dev/pytest-asyncio/issues/68)

```python
@pytest.fixture(scope="session")
def event_loop(request):
    """
    Redefine the event loop to support session/module-scoped fixtures;
    see https://github.com/pytest-dev/pytest-asyncio/issues/68
    """
    policy = asyncio.get_event_loop_policy()
    loop = policy.new_event_loop()

    try:
        yield loop
    finally:
        loop.close()
```

After upgrading to 0.17, our test suite fails with

```python
_______________________________________________________________________________________ ERROR at teardown of test_XXXX_______________________________________________________________________________________

fixturedef = <FixtureDef argname='event_loop' scope='session' baseid='tests'>, request = <SubRequest 'event_loop' for <Function test_secret_settings_are_not_serialized>>

    @pytest.hookimpl(trylast=True)
    def pytest_fixture_post_finalizer(fixturedef, request):
        """Called after fixture teardown"""
        if fixturedef.argname == "event_loop":
            policy = asyncio.get_event_loop_policy()
            # Clean up existing loop to avoid ResourceWarnings
>           policy.get_event_loop().close()

/opt/homebrew/Caskroom/miniconda/base/envs/orion-dev-38/lib/python3.8/site-packages/pytest_asyncio/plugin.py:163: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <asyncio.unix_events._UnixDefaultEventLoopPolicy object at 0x7ff8eaa8f8b0>

    def get_event_loop(self):
        """Get the event loop for the current context.
    
        Returns an instance of EventLoop or raises an exception.
        """
        if (self._local._loop is None and
                not self._local._set_called and
                isinstance(threading.current_thread(), threading._MainThread)):
            self.set_event_loop(self.new_event_loop())
    
        if self._local._loop is None:
>           raise RuntimeError('There is no current event loop in thread %r.'
                               % threading.current_thread().name)
E           RuntimeError: There is no current event loop in thread 'MainThread'.

/opt/homebrew/Caskroom/miniconda/base/envs/orion-dev-38/lib/python3.8/asyncio/events.py:639: RuntimeError
```

Modifying the fixture to set the loop with the policy at the end resolves the error

```python
    try:
        yield loop
    finally:
        loop.close()

    policy.set_event_loop(loop)
```

Leaving the loop open or setting the loop with the policy before yielding does not resolve the error.

I'm not sure why this is happening now, not sure if it needs a fix here.

We're using `asyncio_mode = auto`.

Thanks for the quick reply.

Here's a repository with an MRE: https://github.com/madkinsz/pytest-asyncio-257

It looks like the issue is calling `asyncio.run(...)` in one of the tests.

e.g.
```python
import asyncio


async def test_an_async_test_is_required():
    """Without an async test, the error will not be raised"""
    pass


def test_run_async_in_new_loop():
    async def foo():
        pass
    asyncio.run(foo())
```

```
❯ pytest . --setup-show
============================================================================================================================ test session starts =============================================================================================================================
platform darwin -- Python 3.8.12, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/mz/dev/contrib/pytest-asyncio-257, configfile: setup.cfg
plugins: env-0.6.2, flaky-3.7.0, anyio-3.4.0, xdist-2.4.0, asyncio-0.17.0, forked-1.3.0
collected 2 items                                                                                                                                                                                                                                                            

test_bug.py 
SETUP    S event_loop
        test_bug.py::test_an_async_test_is_required (fixtures used: event_loop).
        test_bug.py::test_run_async_in_new_loop.
TEARDOWN S event_loopE

=================================================================================================================================== ERRORS ===================================================================================================================================
______________________________________________________________________________________________________________ ERROR at teardown of test_run_async_in_new_loop _______________________________________________________________________________________________________________

fixturedef = <FixtureDef argname='event_loop' scope='session' baseid=''>, request = <SubRequest 'event_loop' for <Function test_an_async_test_is_required>>

    @pytest.hookimpl(trylast=True)
    def pytest_fixture_post_finalizer(fixturedef, request):
        """Called after fixture teardown"""
        if fixturedef.argname == "event_loop":
            policy = asyncio.get_event_loop_policy()
            # Clean up existing loop to avoid ResourceWarnings
>           policy.get_event_loop().close()

/opt/homebrew/Caskroom/miniconda/base/envs/orion-dev-38/lib/python3.8/site-packages/pytest_asyncio/plugin.py:163: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <asyncio.unix_events._UnixDefaultEventLoopPolicy object at 0x7f934806a8b0>

    def get_event_loop(self):
        """Get the event loop for the current context.
    
        Returns an instance of EventLoop or raises an exception.
        """
        if (self._local._loop is None and
                not self._local._set_called and
                isinstance(threading.current_thread(), threading._MainThread)):
            self.set_event_loop(self.new_event_loop())
    
        if self._local._loop is None:
>           raise RuntimeError('There is no current event loop in thread %r.'
                               % threading.current_thread().name)
E           RuntimeError: There is no current event loop in thread 'MainThread'.

/opt/homebrew/Caskroom/miniconda/base/envs/orion-dev-38/lib/python3.8/asyncio/events.py:639: RuntimeError
========================================================================================================================== short test summary info ===========================================================================================================================
ERROR test_bug.py::test_run_async_in_new_loop - RuntimeError: There is no current event loop in thread 'MainThread'.
========================================================================================================================= 2 passed, 1 error in 0.10s =========================================================================================================================

--- a/README.rst
+++ b/README.rst
@@ -259,6 +259,7 @@ Changelog
 0.17.1 (UNRELEASED)
 ~~~~~~~~~~~~~~~~~~~
 - Fixes a bug that prevents async Hypothesis tests from working without explicit ``asyncio`` marker when ``--asyncio-mode=auto`` is set. `#258 <https://github.com/pytest-dev/pytest-asyncio/issues/258>`_
+- Fixed a bug that closes the default event loop if the loop doesn't exist `#258 <https://github.com/pytest-dev/pytest-asyncio/issues/257>`_
 
 0.17.0 (22-01-13)
 ~~~~~~~~~~~~~~~~~~~

--- a/README.rst
+++ b/README.rst
@@ -259,7 +259,7 @@ Changelog
 0.17.1 (UNRELEASED)
 ~~~~~~~~~~~~~~~~~~~
 - Fixes a bug that prevents async Hypothesis tests from working without explicit ``asyncio`` marker when ``--asyncio-mode=auto`` is set. `#258 <https://github.com/pytest-dev/pytest-asyncio/issues/258>`_
-- Fixed a bug that closes the default event loop if the loop doesn't exist `#258 <https://github.com/pytest-dev/pytest-asyncio/issues/257>`_
+- Fixed a bug that closes the default event loop if the loop doesn't exist `#257 <https://github.com/pytest-dev/pytest-asyncio/issues/257>`_
 
 0.17.0 (22-01-13)
 ~~~~~~~~~~~~~~~~~~~

--- a/pytest_asyncio/plugin.py
+++ b/pytest_asyncio/plugin.py
@@ -169,8 +169,13 @@ def pytest_fixture_post_finalizer(fixturedef, request):
     """Called after fixture teardown"""
     if fixturedef.argname == "event_loop":
         policy = asyncio.get_event_loop_policy()
-        # Clean up existing loop to avoid ResourceWarnings
-        policy.get_event_loop().close()
+        try:
+            loop = policy.get_event_loop()
+        except RuntimeError:
+            loop = None
+        if loop is not None:
+            # Clean up existing loop to avoid ResourceWarnings
+            loop.close()
         new_loop = policy.new_event_loop()  # Replace existing event loop
         # Ensure subsequent calls to get_event_loop() succeed
         policy.set_event_loop(new_loop)

--- a/tests/test_event_loop_scope.py
+++ b/tests/test_event_loop_scope.py
@@ -29,3 +29,9 @@ def test_3():
     current_loop = asyncio.get_event_loop_policy().get_event_loop()
     # Now the event loop from test_2 should have been cleaned up
     assert loop is not current_loop
+
+
+def test_4(event_loop):
+    # If a test sets the loop to None -- pytest_fixture_post_finalizer()
+    # still should work
+    asyncio.get_event_loop_policy().set_event_loop(None)

